pkg/docker/config.go

```diff
@@ -3,47 +3,120 @@ package docker
 import (
 	"encoding/base64"
 	"encoding/json"
+	"fmt"
+	"net/url"
 	"strings"
+
+	"github.com/google/go-containerregistry/pkg/name"
 )
 
-// ServerCredentials represent credentials used to login to a Docker server.
-type ServerCredentials struct {
-	Auth     string `json:"auth"`
-	Username string `json:"username"`
-	Password string `json:"password"`
+type BasicAuth string
+
+func NewBasicAuth(username, password string) BasicAuth {
+	var v = new(BasicAuth)
+	v.Encode(username, password)
+	return *v
 }
 
-// Credentials represents Docker credentials which are typically stored in `~/.docker/config.json`.
-type Credentials struct {
-	Auths map[string]ServerCredentials `json:"auths"`
+func (v *BasicAuth) Encode(username, password string) {
+	*v = BasicAuth(base64.StdEncoding.EncodeToString(
+		[]byte(fmt.Sprintf("%s:%s", username, password))))
 }
 
-func ReadCredentialsFromBytes(contents []byte) (cfg map[string]ServerCredentials, err error) {
-	var credentials Credentials
-	if err = json.Unmarshal(contents, &credentials); err != nil {
-		return nil, err
+func (v *BasicAuth) Decode() (username, password string, err error) {
+	bytes, err := base64.StdEncoding.DecodeString(string(*v))
+	if err != nil {
+		return
 	}
-	return encodeAuth(credentials.Auths)
+	split := strings.Split(string(bytes), ":")
+
+	username = split[0]
+	password = split[1]
+	return
+}
+
+func (v BasicAuth) String() string {
+	return "[REDACTED]"
+}
+
+// Auth represent credentials used to login to a Docker registry.
+type Auth struct {
+	Auth     BasicAuth `json:"auth,omitempty"`
+	Username string    `json:"username,omitempty"`
+	Password string    `json:"password,omitempty"`
+}
+
+func (v Auth) String() string {
+	return "[REDACTED]"
+}
+
+// Config represents Docker configuration which is typically saved as `~/.docker/config.json`.
+type Config struct {
+	Auths map[string]Auth `json:"auths"`
+}
+
+func (c *Config) Read(contents []byte) (err error) {
+	if err := json.Unmarshal(contents, c); err != nil {
+		return err
+	}
+	c.Auths, err = decodeAuth(c.Auths)
+	return
 }
 
-func encodeAuth(config map[string]ServerCredentials) (encodedConfig map[string]ServerCredentials, err error) {
-	encodedConfig = make(map[string]ServerCredentials)
+func decodeAuth(config map[string]Auth) (encodedConfig map[string]Auth, err error) {
+	encodedConfig = make(map[string]Auth)
```

Comment:
It looks `decodedConfig` to me.

Reply:
Yup. Good catch. Here we decode credentials.